Battery strategy 2

PastCoder opened the pull request:
This time a little bigger change - actually it is not a change but an addition. 

I propose an alternative strategy when to switch on/off the KOSTAL Smart Battery Management. The main information about the strategy is added in Readme.md and Readme_de.md. I've added there a section "Smart battery control". 
The new strategy is added as option which can be selected in the adapter settings. The default strategy is the previous strategy. 

 (!) There is one change in the settings compare to the current github version: I changed the setting for the feed-in limitation from a percentage to an absolute value. It give a little more flexibility (e.g. if the actually panel efficiency is different from the nominal one). 

I am using this strategy since a while now and it seems to work quite well. 

 (!) There is one thing which I did not manage: When you open the adapter settings, it shown always strategy 1 as current selection. I did nit figute out how to set the dropdown list to the previously select value when loading the settings. 

When transferring the implementation from my experimental branch to this proposed one, I split the changes in three pull request: 
- Adjustment of the settings
- Extract code into the function setSmartBatteryControl() to make the code a little more readable 
- Addition of the strategy 2

Test usual: Additional testing and feedback welcome :-)  I am using strategy 2 so that this one wil lbe tested anyhow (at least with my setup).

StrathCole commented on admin/index_m.html:
```diff
@@ -124,7 +124,7 @@
 					if (id === 'password') {
 						value = encrypt(secret, value);
-					} else if(id === 'min_minsoc' || id === 'max_minsoc' || id === 'enable_bm_minsoc' || id === 'port') {
+					} else if(id === 'min_minsoc' || id === 'max_minsoc' || id === 'enable_bm_minsoc' || id === 'port'  || id === 'max_feed_in_power') {
```
I think you will need to do a conversion here if there is a "max_feed_in_percentage" setting on loading the settings. Otherwise everyone updating and opening the settings will have their values deleted and the default of 5000 there.

Maybe you could check if obj['max_feed_in_percentage'] is set and max_feed_in_power is 5000 or unset and then calc the setting by using the pv power * percentage.

StrathCole commented on lib/plenticore.js:
```diff
@@ -808,10 +811,16 @@ function getPanelWp() {
-function getMaxFeedInFactor() {
-	let factor = adapter.config.max_feed_in_percentage / 100  // translate percentage to factor
+function getMaxFeedInPower() {
+	let power = adapter.config.max_feed_in_power;
```
This might lead to a problem when someone updates the adapter and does not update the config (which cannot be expected to be done on minor version changes).

StrathCole reviewed (commented):
Thank you for your contribution(s). Great work.

PastCoder commented:
I have added code to migrate from old settings (either with max_feed_in_percentage or without) to the new settings in index_m.html and plenticore.js.
But as I don't know how I can remove existing values from the adapter config, I could not test the migration. I need to rely on a hint from you how to do that or to rely on your test/review.

Do you know how the make it work that the strategy dropdown updates to the actual setting when load the setting page?

StrathCole commented:
> Do you know how the make it work that the strategy dropdown updates to the actual setting when load the setting page?

In index_m.html around line 47 there is a check for the input type. I think you need to loop through the corresponding option elements in the form and then add "selected" to that one matching the settings value.

StrathCole commented:
> I have added code to migrate from old settings (either with max_feed_in_percentage or without) to the new settings in index_m.html and plenticore.js. But as I don't know how I can remove existing values from the adapter config, I could not test the migration. I need to rely on a hint from you how to do that or to rely on your test/review.
> 
> Do you know how the make it work that the strategy dropdown updates to the actual setting when load the setting page?

In index_m.html from line 127 (the save handler) you could just add something like obj['max_feed_in_percentage'] = null, I think.

PastCoder commented:
> In index_m.html around line 47 there is a check for the input type. I think you need to loop through the corresponding option elements in the form and then add "selected" to that one matching the settings value.

I found a way how it works. Programming the logic is easy but as I am not a Javascript developer some of the stuff around is hard :-)

PastCoder commented:
> add something like obj['max_feed_in_percentage'] = null

I've added this as proposed. Because I am not very familiar with Javascript I am not sure if the entry is now really gone or if it now jus has the value null. To be on the safe side, I added a check for null in getMaxFeedInPower(). By this I found a mistake there.

PastCoder commented:
Is in index_m.html the function fillInstances() still used? Or could this be removed to keep the code clean?

StrathCole commented:
> Is in index_m.html the function fillInstances() still used? Or could this be removed to keep the code clean?

I think it was just for the weather forecast adapter selection and should no longer be used.

StrathCole commented on admin/index_m.html:
```diff
@@ -336,14 +357,22 @@
 				</div>
 				<div class="col s6 input-field">
 					<input type="number" class="value" id="min_minsoc" />
-					<label for="min_minsoc" class="translate">minimum minsoc</label>
```
Have you deleted the "for" by accident? It states which input the label is for.

PastCoder replied:
No, that was a pure mistake. I have re-added it again. Thanks for seeing this!

StrathCole commented:
> I've added this as proposed. Because I am not very familiar with Javascript I am not sure if the entry is now really gone or if it now jus has the value null. To be on the safe side, I added a check for null in getMaxFeedInPower(). By this I found a mistake there.

Looks good to me how you solved it :+1:

PastCoder commented:
> I think it was just for the weather forecast adapter selection and should no longer be used.

I have removed it. Just as a proposal you can decide on.

StrathCole commented:
Thank you. I will issue a new release.
